LSP Selection Ranges

ChrisPenner opened the pull request:
## Overview

For https://github.com/unisonweb/unison/issues/4223

## Implementation notes

How does it accomplish it, in broad strokes? i.e. How does it change the Haskell codebase?

## Interesting/controversial decisions

Include anything that you thought twice about, debated, chose arbitrarily, etc.
What could have been done differently, but wasn't? And why?

## Test coverage

Have you included tests (which could be a transcript) for this change, or is it somehow covered by existing tests?

Would you recommend improving the test coverage (either as part of this PR or as a separate issue) or do you think it’s adequate?

If you only tested by hand, because that's all that's practical to do for this change, mention that.

## Loose ends

Link to related issues that address things you didn't get to. Stuff you encountered on the way and decided not to include in this PR.

sellout committed:
Extract the `Doc` lexer into a top-level function

sellout committed:
Separate the `Doc` lexer from the Unison lexer

`doc2` is a Unison lexer that traverses a `Doc`.

`docBody` is the actual `Doc` lexer that is ignorant of the fact that Unison wraps `Doc` blocks in `{{`/`}}`.

sellout committed:
Un-hiding the `Doc` parser

`doc2` was a parser in lexer’s clothing. It would parse recursively, but
then return the result as a flat list of tokens.

This separates the parsing from the “unparsing” (which returns the
tokens), so now we have a parser to a recursive `Doc` structure. This
currently immediately applies the unparser, and should result in an
identical stream of tokens as the previous version. Eventually, we
should be able to avoid unparsing the `Doc` structure.

ChrisPenner committed:
Fix behaviour of branch command

ChrisPenner committed:
Un-ignore transcript-parser-commands.md and fix it.

ChrisPenner committed:
automatically run ormolu

mitchellwrosen committed:
make decl coherency check more abstract to support "get all violations" use case

ChrisPenner committed:
Mostly Fix up deleting the branch you're on

ChrisPenner committed:
Merge trunk

ChrisPenner committed:
Include trailing '!' in identifier names for hover

ChrisPenner committed:
automatically run ormolu

aryairani committed:
Merge pull request #5175 from unisonweb/cp/trailing-bang-lsp

aryairani committed:
Merge pull request #5172 from unisonweb/cp/more-project-transcripts

mitchellwrosen committed:
report constructor aliases in `todo`

mitchellwrosen committed:
report missing constructor names in `todo`

mitchellwrosen committed:
report nested decl aliases in `todo`

mitchellwrosen committed:
report stray constructors in `todo`

mitchellwrosen committed:
⅄ trunk → 24-07-01-todo-merge-precondition-violations

ChrisPenner committed:
Fix behaviour of deleting last project or branch

ChrisPenner committed:
Deprecate root reflog behaviour in favour of project/branch reflogs

ChrisPenner committed:
Implement DB combinators for project and branch reflogs

sellout committed:
Move the `Annotated` class to the `Ann` module

This is in preparation for using `Ann` in the `Lexer` module, as that module
actually does some parsing.

ChrisPenner committed:
More WIP on reflog commands

ChrisPenner committed:
Specialize 'fromHash' to CausalHash

ChrisPenner committed:
Add reflog.global command

ChrisPenner committed:
Update reflog command WIP

aryairani committed:
move pretty-print helpers out of CommandLine.hs

to avoid a potential cycle between CommandLine.hs and InputPatterns.hs
that otherwise came up on another branch.

InputPatterns.hs also has formatting helpers that could be used in CommandLine.hs.
Maybe they should be moved too, but I haven't thought about it.

aryairani committed:
switch `Left` to `Right` on help outputs; add draft transcripts

sellout committed:
Fix shared CLI error message

- don’t add an extra indent level to the entire output
- do add a `warnCallout` to the message
- don’t re-wrap formatted error output
- remove the accidentally-included entire `help <command>` output

sellout committed:
Make CLI error messages more consistent

- remove `warnCallout` and `warn` from individual messages
- make sure no error messages return their `help` content
- add some documentation about how to write failure messages

sellout committed:
switch `Left` to `Right` on help outputs; add draft transcripts (#1)

Maybe some duplication/conflicts here unintentionally.

The big transcript is just a WIP, we could make it prettier.

sellout committed:
Fix formatting of an error message

It was leaving out spaces between literal text and generated text.

sellout committed:
Update transcript outputs

ChrisPenner committed:
Improve LSP completion sorting

sellout committed:
Merge remote-tracking branch 'upstream/trunk' into better-CLI-error-messages

sellout committed:
Remove a help message from an error

This silpped by in the merge.

sellout committed:
Remove another warning flag from a message

Another one missed in the merge.

ChrisPenner committed:
Add unused binding detection to LSP

aryairani committed:
change a punctuation

aryairani committed:
Merge pull request #5054 from sellout/better-CLI-error-messages

aryairani committed:
Merge pull request #5180 from unisonweb/arya/commandline-helpers

ChrisPenner committed:
Swap to using para instead of cata

aryairani committed:
Merge pull request #5182 from unisonweb/cp/improve-lsp-completion

sellout committed:
Don’t “un-parse” `Doc`.

This removes the layer that makes the `Doc` parser look like a lexer and
replaces it with a function that converts the Doc structure directly
Unison Terms.

ChrisPenner committed:
Revert back to simple cata

ChrisPenner committed:
Move warnings to the top of the term

aryairani committed:
fix compile error and use `makeExample` more

ChrisPenner committed:
Compress unused bindings into a single diagnostic

ChrisPenner committed:
automatically run ormolu

aryairani committed:
Merge pull request #5185 from unisonweb/arya/commandline-helpers

ChrisPenner committed:
Fix diagnostic still reporting when no unused bindings

aryairani committed:
Merge pull request #5186 from unisonweb/lsp/unused-binding-detection

ChrisPenner committed:
Merge trunk

ChrisPenner committed:
Fix up reset transcripts for projects

ChrisPenner committed:
Rerun transcripts

ChrisPenner committed:
Disable view.global transcript for now

ChrisPenner committed:
disable names.global transcript for now

ChrisPenner committed:
automatically run ormolu

ChrisPenner committed:
Get tests building again.

ChrisPenner committed:
Fix up round-trip tests

ChrisPenner committed:
Fix integration-tests

ChrisPenner committed:
Merge pull request #5188 from unisonweb/lsp/unused-binding-detection

Unused binding detection cleanup

aryairani committed:
Merge pull request #5155 from sellout/cp/ghc-upgrade

aryairani committed:
avoid numeric vs textual clash in args length error

`numerals` doesn't seem to be maintained anymore, however

ChrisPenner committed:
Merge trunk, clean up random extra files

ChrisPenner committed:
Improve docs

ChrisPenner committed:
Add undo.md transcript

ChrisPenner committed:
Use annotations from Abs instead

ChrisPenner committed:
Attempt to fix Unused Binding locations

mitchellwrosen committed:
synhash var bugfix: hash debruijn indexes, not var names

also adds a debug.synhash.term command for debugging

mitchellwrosen committed:
automatically run ormolu

mitchellwrosen committed:
delete unused import

aryairani committed:
Update unison-cli/src/Unison/CommandLine/InputPatterns.hs

Co-authored-by: Greg Pfeil <greg@technomadic.org>

ChrisPenner committed:
Update LSP to listen for changes

ChrisPenner committed:
Ignore more scratchfiles

ChrisPenner committed:
Write new Transactional Signal type

ChrisPenner committed:
No in-memory branch in loop-state

mitchellwrosen committed:
add transcript

aryairani committed:
Merge pull request #5192 from unisonweb/topic/numerals

aryairani committed:
Merge pull request #5194 from unisonweb/synhash-bugfix

pchiusano committed:
Don't use DelayForceChar as syntax style for `blah.default()`

sellout committed:
Pin Haskell tool versions for VS Code

This also has the flake get its version pins (when possible) from the VS
Code settings. And we pin Cabal now, too.

pchiusano committed:
Merge pull request #5195 from unisonweb/force-highlighting

Use `SyntaxText.Unit` as style for `foo()`

ChrisPenner committed:
Add isTranscriptTest to Cli.Env

ChrisPenner committed:
Omit times from project reflogs

ChrisPenner committed:
Merge trunk

ChrisPenner committed:
preload branches into the branch cache when switching projects

mitchellwrosen committed:
adjust some output

ChrisPenner committed:
Fix test build

mitchellwrosen committed:
better rendering of conflicted names in `todo`

sellout committed:
Replace transcript parser with `cmark`

We don’t need a very rich parser for transcripts, but we _do_ need to
reliably identify fenced code blocks, and that implies a number of
subtle cases. Using a polished CommonMark parser/printer handles those
subtleties for us.

I chose `cmark` for a few reasons:
- it’s a wrapper around `libcmark`, which is the reference
  implementation of CommonMark, so it should be correct;
- it provides both a parser and a printer (unlike MMark); and
- it is extremely fast (about 20x faster than MMark), so the fact that
  our home-rolled parser got to skip over everything that’s not a block
  isn’t an issue.).

This only _partially_ uses the `cmark` printer. I think it should use it
entirely, but for the cases where we do streaming output (processing UCM
commands, etc.) it’s a more involved change. So I think it should be
handled separately.

sellout committed:
Update the transcripts with `cmark`

`cmark`’s pretty-printer matches our output pretty well, with a few differences:
- it puts a space between the fence and the info string for in code blocks;
- it prefers `-` over `*` for bulleted lists (as do I) and it indents them;
- it `\`-escapes certain chars very conservatively;
- it prefers indented/unfenced code blocks if there is no info string; and
- it prefers `*` over `_` (unlike any sane person).

This also shows how the change fixes a number of issues:
- fix2158-1.output.md also illustrates how this change fixes #1809;
- alias-many.output.md and input-parse-errors.output.md show how fenced
  code blocks without an info string would use the beginning of the
  content as the info string;
- transcripts-round-trip/main.output.md shows how output blocks for
  generated `unison` stanzas (which could contain nested fenced blocks)
  might not have long-enough fences; and
- error-messages.output.md and generic-parse-errors.output.md show how
  Unison errors were reported on the wrong line number (and thus the
  printed error lines were also incorrect).

sellout committed:
Add a space before code block info strings

This is for consistency with the `cmark` style. Now the blocks we still
pretty-print ourselves will match the bulk of them that `cmark`
produces.

sellout committed:
Fix a few transcripts with incorrect Markdown

These weren’t errors in any way, but the `cmark`-produced outputs made
it clear that some of our transcripts weren’t formatted the way we
intended.

ChrisPenner committed:
Revive move-branch confirmations

sellout committed:
Add transcripts for fixed issues

This shows that this change fixes #1421, #2826, #5141, and #5168.

ChrisPenner committed:
Docs and transcripts

aryairani committed:
Merge pull request #5177 from unisonweb/24-07-01-todo-merge-precondition-violations

ChrisPenner committed:
Fix move-namespace transcript

ChrisPenner committed:
Fix merge transcript

ChrisPenner committed:
Fix names transcript

ChrisPenner committed:
Fix Upgrade

ChrisPenner committed:
Typo

sellout committed:
Terminate API code blocks correctly

They were missing trailing newlines, so as you can see in the diff, some
transcripts had invalid output.

sellout committed:
Merge remote-tracking branch 'upstream/trunk' into new-transcript-parser

ChrisPenner committed:
Attempt to fix up update.old

sellout committed:
Update a couple transcripts to the new formatting

These aren‘t run as part of ./scripts/check.sh, so they got missed in the initial pass.

ChrisPenner committed:
No leading dot on brps

aryairani committed:
Merge pull request #5204 from sellout/new-transcript-parser

aryairani committed:
Merge branch 'trunk' into cp/project-root

# Conflicts:
#	unison-cli-integration/integration-tests/IntegrationTests/transcript.output.md
#	unison-src/transcripts-round-trip/main.output.md
#	unison-src/transcripts/add-run.output.md
#	unison-src/transcripts/bug-strange-closure.output.md
#	unison-src/transcripts/cycle-update-5.output.md
#	unison-src/transcripts/delete.output.md
#	unison-src/transcripts/diff-namespace.output.md
#	unison-src/transcripts/move-namespace.output.md
#	unison-src/transcripts/name-selection.output.md
#	unison-src/transcripts/names.output.md
#	unison-src/transcripts/namespace-dependencies.output.md
#	unison-src/transcripts/propagate.output.md
#	unison-src/transcripts/reflog.output.md
#	unison-src/transcripts/reset.output.md
#	unison-src/transcripts/tab-completion.output.md
#	unison-src/transcripts/transcript-parser-commands.output.md

ChrisPenner committed:
Use a branch root semispace cache

ChrisPenner committed:
Remerge remote

ChrisPenner committed:
Merge trunk

dolio committed:
Use pretty text in MsgTrace case of debug toText

ChrisPenner committed:
Swap branch-relative-paths back to relative

ChrisPenner committed:
Fix diff-helper to use the names from both branches

ChrisPenner committed:
Port fix2254 to new update

aryairani committed:
Merge pull request #5210 from unisonweb/topic/msg-trace

mergify[bot] committed:
Merge pull request #5041 from unisonweb/cp/project-root

Project Roots

sellout committed:
Refactoring TranscriptParser

A bunch of small changes in TranscriptParser

- remove dead code
- don’t use `Show` for formatted output
- put processed blocks in a separate sum type from unprocessed blocks
- remove `Transcript` from identifiers (changed importers to use
 `qualified as Transcript`)
- deduplicated some error reporting

And one happy fix, IMO – got rid of the `Text.init` that plagued me in

dolio committed:
Use builtin reference names in foreign decompiling

E.G. print <Promise> instead of <Foreign>

mergify[bot] committed:
Merge pull request #5216 from unisonweb/topic/decomp-foreign

Use builtin reference names in foreign decompiling

ChrisPenner committed:
Compiling with lamda annotations

ChrisPenner committed:
Add unused binding test

ChrisPenner committed:
Merge trunk

ChrisPenner committed:
Keep most recent path and branches through migration to version 17

ChrisPenner committed:
Reset user to project root if they were on a sub-namespace.

aryairani committed:
Merge pull request #5218 from unisonweb/lsp/fix-unused-binding-locs

aryairani committed:
Merge pull request #5220 from unisonweb/cp/keep-current-branch

neduard committed:
update command for integration tests in docs

neduard committed:
create gitignore in scheme-libs/racket/unison/

neduard committed:
update scheme-libs/racket/unison/Readme.md

neduard committed:
use libb2 for all blake functions

required if compiling on systems that use eg libressl

neduard committed:
use /usr/bin/env sh for jit-tests.sh

this is more portable and ensures we can compile on systems without bash (eg *BSD)

ChrisPenner committed:
Don't evaluate the unison file if there are no watch expressions.

neduard committed:
fix warning: add missing Functor

neduard committed:
Merge remote-tracking branch 'origin/trunk' into merged-ghc-upgrade-try-2

aryairani committed:
Merge pull request #5223 from unisonweb/cp/skip-eval

neduard committed:
update transcripts

ChrisPenner committed:
Better recursive ref search

aryairani committed:
Merge pull request #5226 from unisonweb/cp/faster-dep-collection

aryairani committed:
Merge pull request #5224 from neduard/merged-ghc-upgrade-try-2

sellout committed:
Avoid duplicating YAML info in Markdown

Reviewing the last merge into #5142, I noticed some duplicated and out-
of-date information. This brings things up-to-date and slightly reduces
the duplication.

- bumped Ormolu used by ci.yaml from 0.5.2.0 to 0.7.2.0, to match
  flake.nix
- removed Markdown that claimed ci.yaml was using Ormolu 0.5.0.1
- moved description from `base-codebase` in Markdown to comment on
 `runtime_tests_codebase` in ci.yaml (and updated it to refer to
   builtin-tests/interpreter-tests.md instead of builtin-tests/base.md)
- removed `unison_src_test_results` as it’s no longer managed as a
  single variable
- moved other comments from Markdown to ci.yaml
- added Markdown recommending to look in ci.yaml for specifics
- rearranges the order of vars in ci.yaml to match the order they were
  presented in Markdown

One thing I wasn’t sure how to map over: Markdown claims Racket 8.7 is
used in CI, but ci.yaml doesn’t mention any Racket version.

neduard committed:
fix jit generation in ci.yaml

sellout committed:
Don’t let Stack tell contributors to upgrade

We use a specific version of Stack, we don’t want Stack complaining to
contributors that it doesn’t happen to be the latest.

ChrisPenner committed:
Remove now-unused sync types

sellout committed:
Don’t let Stack complain about Nix

We intentionally disable Stack’s Nix integration when it’s run from a
Nix shell. We don’t want Stack complaining that we should enable it
since it found `nix` on the path.

sellout committed:
Split `TranscriptParser` into three modules

- `Transcript` – the data model
- `Transcript.Parser` – the parser and printer
- `Transcript.Runner` – the runner

There is unfortunately still some printing tightly coupled to the
runner, but this makes it more obvious. Also, the runner is the only
component tied to the CLI.

aryairani committed:
Merge pull request #5228 from sellout/pin-vscode-versions

aryairani committed:
Merge pull request #5231 from sellout/dont-upgrade-stack

aryairani committed:
Merge pull request #5227 from sellout/deduplication

mitchellwrosen committed:
automatically run ormolu

sellout committed:
Document Nix cache configuration

sellout committed:
Add docs for updating the Nix config after editing it

sellout committed:
Minor formatting cleanup

So my editor stops fixing things for me.

sellout committed:
Clean up devShell definition

Mostly formatting, but also
- removes Apple’s Cocoa from the dependencies,
- moves some dependencies from `buildInputs` to `nativeBuildInputs`.

sellout committed:
Add `cachix` to the devShell

This tool manages the Unison cache, so it’s useful to provide in dev
environments. Actually pushing to the cache requires a secret key.

ChrisPenner committed:
Merge trunk

ChrisPenner committed:
automatically run ormolu

neduard committed:
just whitespace changes

neduard committed:
update transcripts-manual/gen-racket-libs.output.md

neduard committed:
minor fixups

neduard committed:
remove extra word and mention libressl

mergify[bot] committed:
Merge pull request #5237 from neduard/jit-support-non-gnu-unix-systems

Support native compilation on non-GNU Unix systems

sellout committed:
Run Ormolu over the entire codebase

With #5142, Ormolu was upgraded from 0.5.2.0 to 0.7.2.0. This formats
the codebase to avoid spurious formatting comingled in other commits.

Almost all of the changes are simply wrapping single constraints in parens, like
```diff
-hashBranch :: forall m. Monad m => Branch m -> m BranchHash
+hashBranch :: forall m. (Monad m) => Branch m -> m BranchHash
```

There is also some reordering of language pragmas and imports,
indentation correction (some of which gets precedence wrong), and switching
some Haddock from `-- ^` to `-- |` .

sellout committed:
Manually reformat chains of single constraints

Changing `(A x) => (B y) => x -> y` to `(A x, B y) => x -> y`.

sellout committed:
Inform Ormolu of some operator precedences

This adds a .ormolu file to tell Ormolu what some operator precedences
are, to improve indentation of multi-line operator sequences.

aryairani committed:
Merge pull request #5239 from sellout/reformatting

aryairani committed:
Merge pull request #5230 from unisonweb/cp/unused-sync-types

aryairani committed:
Merge pull request #5233 from sellout/cachix-in-dev-shell

aryairani committed:
Merge pull request #5235 from sellout/refactor-transcript-parser

aryairani committed:
Merge pull request #5232 from sellout/doc-nix-cache

ceedubs committed:
Don't limit max line length on .u scratch files

The recently added `.editorconfig` sets a max line length of 120 characters for all file types. I found this to be
pretty frustrating when writing Unison scratch files. While writing Unison code my editor would seemingly randomly
create a new line, generating invalid Unison code. This came up more often when writing `Doc` values, since I tend to
write longer lines in prose than in code.

The formatting in scratch files is ephemeral, so I don't think that there is any good reason to force a max line length
on it.

ChrisPenner committed:
Remove reset-root, fix docs for reset

ChrisPenner committed:
automatically run ormolu

ChrisPenner committed:
Update transcripts

aryairani committed:
Merge pull request #5243 from unisonweb/cp/remove-reset-root

Remove reset-root, fix docs for reset

sellout committed:
Extract `preParse` from `lexer`

After running the core of the lexer, the `lexer` function then does some
work to turn the stream into a tree, and reorder some lexemes. It then
throws away the tree structure.

This is the first step of preserving the tree structure for the parser.
It extracts the “pre-parser” from `lexer` so
that it can eventually be used _after_ the lexer, rather than internally.

This also moves `fixup` to be applied on each block as we reorder it,
rather than across the entire stream at the end (since the goal is to
not _have_ an entire stream any more).

ChrisPenner committed:
branch.reflog -> reflog

aryairani committed:
Merge pull request #5240 from neduard/jit-ci-fix

mitchellwrosen committed:
update work

sellout committed:
Expose `preParse` to the parser

sellout committed:
Rename `T` to `BlockTree`

ChrisPenner committed:
Merge pull request #5244 from unisonweb/cp/reflog-name

branch.reflog -> reflog

mitchellwrosen committed:
simplify update

mitchellwrosen committed:
get transcripts passing

sellout committed:
Restructure `BlockTree`

We now build the stanzas at the same time as the tree, and don’t discard them after reordering.

This also changes the closing element of `Block` to be `Maybe` instead of `[]`.

mitchellwrosen committed:
move some code around

mitchellwrosen committed:
⅄ trunk → update-defn-order

mitchellwrosen committed:
simplify checkDeclCoherency again

ceedubs committed:
Merge pull request #5242 from unisonweb/editor-config-u

Don't limit max line length on .u scratch files

mitchellwrosen committed:
delete unused imports

sellout committed:
Add some transcripts that should error, but don’t

sellout committed:
Ensure transcript parser consumes entire stanzas

With the switch to `cmark`, the “second phase” parsing of individual
stanzas omitted an EOF check to ensure that the entire stanza had been
parsed. This resulted in parses where we end up with truncated sets of
UCM commands or API requests, which could either result in premature
success or failures occurring later in the transcript, where they’d
complain about the wrong thing.

sellout committed:
Add transcripts for some already-fixed issues

This fixes #1327 (which was actually fixed by #5056 or #5061) and #3977
was already closed, but without the transcript to avoid regression.

sellout committed:
Allow EOF to close layout blocks

This removes the need to pad the lexer stream with trailing `Close` lexemes. If
EOF is reached, the parser will automatically close any layout blocks (but not
context-free blocks).

sellout committed:
Make comments into Haddock

sellout committed:
Remove unnecessary `docOpen` in Doc parser

aryairani committed:
Merge pull request #5251 from sellout/transcript-consume-entire-stanza

sellout committed:
Split `Doc` into its own module

mitchellwrosen committed:
write pure version of synhashLcaDefns

mitchellwrosen committed:
write pure version of synhashDefns

aryairani committed:
Merge pull request #5238 from neduard/jit-update-readme

mitchellwrosen committed:
write pure version of nameBasedNamespaceDiff

mitchellwrosen committed:
write pure version of checkDeclCoherency

mitchellwrosen committed:
write pure version of lenientCheckDeclCoherency

sellout committed:
Add `Data.Functor.Classes` instances

These are needed for the new Doc types, but had been stubbed out. Moving
the Doc types to their own module forced the changes that got in the way
of generating these with Template Haskell.

sellout committed:
Split Doc parser from Unison lexer

sellout committed:
Simplify `restoreStack`

It’s only used inside `local`, so its attempts to restore the layout are for naught.

sellout committed:
Split the Doc parser into multiple functions

In general, they map to the constructors of the Doc types, with some wiggle room for now.

It’s probably beneficial to review this commit by ignoring whitespace.

ChrisPenner committed:
Merge branch 'trunk' into cp/defn-search-sync

aryairani committed:
Merge pull request #5254 from unisonweb/cp/defn-search-sync

Add FromJSON for Display Object for definition search

sellout committed:
Generalize the Doc parser

It is now completely[^1] independent of the Unison language. The parser takes a few parsers as arguments: one for
identifiers, one for code, and one to indicate the end of the Doc block.

[^1]: There is one last bit to be removed in the next commit – Doc still looks for `type` or `ability` to identify type
links.

sellout committed:
Caught a hardcoded `}}` in the Doc parser

The Doc parser shouldn’t know how Unison terminates Doc blocks.

sellout committed:
Make Doc parser ignorant of type/term distinctions

This was the last thing tying Doc to Unison.

mitchellwrosen committed:
⅄ update-defn-order → 24-07-24-diff-api

mitchellwrosen committed:
make a Unison.Merge module that re-exports most of the merge API

sellout committed:
Simplify Doc parser from `State` to `Reader`

tstat committed:
permit empty matches

tstat committed:
remove unused arg

aryairani committed:
Merge pull request #5256 from unisonweb/travis/empty-pattern-match

mitchellwrosen committed:
break up libdeps merge into diff and apply steps

clojj commented:
@ChrisPenner 
This would be very welcomed!
My editing (coming from other languages) is centered around semantically selecting element (blocks of elements)

ChrisPenner committed:
Add ability to find over EVERY branch.

ChrisPenner committed:
re-add names.global

ChrisPenner committed:
Rerun transcripts

ChrisPenner committed:
Fix transcripts

ChrisPenner committed:
automatically run ormolu

sellout committed:
Add some description to the new transcripts

aryairani committed:
tweak output messages

aryairani committed:
Merge pull request #5258 from unisonweb/cp/global-search

sellout committed:
Add a transcript showing that #5076 was fixed

Some handling of blocks without final newlines was improved in the course of this PR.

Fixes #5076.

sellout committed:
Merge remote-tracking branch 'upstream/trunk' into doc-lexer

aryairani committed:
Merge pull request #5262 from unisonweb/update-defn-order-tweak

ChrisPenner committed:
Fix unused-binding-detection in case patterns

aryairani committed:
debug nix-dev-cache

aryairani committed:
Update nix-dev-cache.yaml

aryairani committed:
Update nix-dev-cache.yaml

aryairani committed:
Update nix-dev-cache.yaml

aryairani committed:
Update .github/workflows/nix-dev-cache.yaml

Co-authored-by: Greg Pfeil <greg@technomadic.org>

aryairani committed:
Update .github/workflows/nix-dev-cache.yaml

ChrisPenner committed:
Merge pull request #5265 from unisonweb/lsp/fix-unused-bindings-in-cases

Fix unused-binding-detection in case patterns

aryairani committed:
Update .github/workflows/nix-dev-cache.yaml

Co-authored-by: Greg Pfeil <greg@technomadic.org>

aryairani committed:
Merge pull request #5264 from unisonweb/nix-tmate

aryairani committed:
Merge pull request #5250 from unisonweb/update-defn-order

aryairani committed:
Merge pull request #5187 from sellout/doc-lexer

aryairani committed:
Merge pull request #5252 from sellout/test-1327

mitchellwrosen committed:
⅄ trunk → 24-07-24-diff-api

aryairani committed:
Merge pull request #5257 from unisonweb/24-07-24-diff-api

ChrisPenner committed:
Merge trunk

ChrisPenner committed:
Attempt to clean up selection ranges with mixed results
